complex/fi_ubertest: add atomic testing

aingerson opened the pull request:
Enable atomic testing in ubertest, including all functions, operations, and datatypes.

Signed-off-by: aingerson <alexia.ingerson@intel.com>

a-ilango commented on complex/ft_test.c:
```diff
@@ -205,15 +218,10 @@ static void ft_format_iov_random(struct iovec *iov, size_t cnt, char *buf,
-			 * least a length of 1 for every IOV.
-			 */
-			weight = (rand() % (len - (cnt - i))) + 1;
-		}
```
Isn't this still needed to properly assign weights?

aingerson replied:
The other iov formatter methods allow for an iov to have a count/length of 0 except for this one so when the number of iovs > total elements being transferred, it ends up diving by 0. This change is to avoid a floating point exception by modifying the range it's selecting between. It wasn't a problem before because the iov_count never exceeded the total length of the message, but the iocs work with counts, not lengths, so they end up being a lot smaller than the lengths.

aingerson replied:
@shefty is it ok to have an iov of length 0 or should this avoid it?

shefty replied:
That should be avoided, so providers don't need to check for it.  At least until support is standardized through the API.

shefty commented on common/shared.c:
```diff
@@ -1089,6 +1090,42 @@ int size_to_count(int size)
+			return sizeof(long double complex);
+		default:
+			return 0;
+	}
```
The case statements are indented one tab too far.  But we can drop this function and use the new fi_query_atomic() call to get the size of the datatypes as defined by the provider.  This is needed for long double, which is allowed by the standard to be just about anything.

For long double, the app should also check if its long double size matches that returned by the provider.

shefty commented on complex/fabtest.h:
```diff
@@ -88,6 +97,9 @@ struct ft_xcontrol {
 	uint64_t		remote_cq_data;
+	enum fi_op		op;
+	enum fi_datatype	datatype;
+	size_t			count;
```
gurgle that's a lot of additions.  Please move these into an xcontrol_atomic sub-structure, just to keep them grouped.

shefty commented on complex/fabtest.h:
```diff
@@ -117,6 +129,7 @@ enum {
 	FT_MAX_AV_TYPES		= 3,
 	FT_MAX_PROV_MODES	= 4,
 	FT_MAX_WAIT_OBJ		= 5,
+	FT_MAX_OPS		= 19,
```
Is this trying to capture FI_ATOMIC_OP_LAST?

shefty commented on complex/ft_domain.c:
```diff
@@ -191,14 +192,53 @@ static int ft_setup_xcontrol_bufs(struct ft_xcontrol *ctrl)
+				FT_PRINTERR("fi_mr_reg", ret);
+				return ret;
+			}
+			ctrl->comp_memdesc = fi_mr_desc(ctrl->mr);
```
This function is getting too long.  Please create a new function that initializes the atomic portions

shefty commented on complex/ft_main.c:
```diff
@@ -120,11 +120,77 @@ static char *ft_class_func_str(enum ft_class_function enum_str)
+	default:
+		return "op_unspec";
+	}
+}
```
Use fi_tostr() with FI_TYPE_ATOMIC_OP

shefty commented on complex/ft_msg.c:
```diff
@@ -304,6 +304,183 @@ static int ft_post_send_rma(void)
+			ft_format_iov(ft_tx_ctrl.iov, iov_count, ft_tx_ctrl.buf,
+				ft_tx_ctrl.count);
+			ft_iov_to_ioc(ft_tx_ctrl.iov, ft_tx_ctrl.ioc, iov_count,
+				ft_tx_ctrl.datatype, ft_tx_ctrl.buf);
```
Maybe just add ft_format_ioc, rather than formatting as an iovec, then converting.  (Saying this without having looked at the conversion function yet).

shefty commented on complex/ft_test.c:
```diff
@@ -205,15 +218,10 @@ static void ft_format_iov_random(struct iovec *iov, size_t cnt, char *buf,
-			weight = (rand() % (len - (cnt - i))) + 1;
-		}
+		else
+			weight = rand() % len;	
```
This allows weight to be 0

shefty commented on complex/ft_test.c:
```diff
@@ -251,6 +270,30 @@ void ft_next_iov_cnt(struct ft_xcontrol *ctrl, size_t max_iov_cnt)
+		case FT_FUNC_COMPARE_ATOMICV:
+		default:
+			ret = fi_compare_atomicvalid(ft_tx_ctrl.ep, ft_tx_ctrl.datatype, ft_tx_ctrl.op, count);
+	}
```
please outdent case statements.

shefty commented on complex/ft_test.c:
```diff
@@ -251,6 +270,30 @@ void ft_next_iov_cnt(struct ft_xcontrol *ctrl, size_t max_iov_cnt)
 		ctrl->iov_iter = 0;
 }
 
+int ft_is_valid_atomic_op(size_t *count)
```
The name of the function (is_valid) implies a boolean return (0 = false, 1 = true), but the behavior is the opposite (is_valid = 0 = true).  Maybe switch to using the fi_query_atomic() call, and check that the size is the same as well.  Then rename to ft_query...  Note that a size mismatch doesn't necessarily indicate an error, except for the fixed size types.

shefty commented:
This is a nice addition to the test.  Something else to keep in mind is that support for tagged atomic operations have been defined.  Please consider how that can be added as well.  You don't have to add it, just consider it as part of the design, to help minimize re-work.  I hope to have tagged atomic support in the socket provider within the next couple of weeks.

aingerson commented:
Updated.
- kept datatype_to_size to allow app to to check datatype size against atomic_attr.size
- modified random iov distribution weight to allow len = cnt

shefty commented on complex/ft_test.c:
```diff
@@ -212,7 +242,7 @@ static void ft_format_iov_random(struct iovec *iov, size_t cnt, char *buf,
 			 * least a length of 1 for every IOV.
 			 */
-			weight = (rand() % (len - (cnt - i))) + 1;
+			weight = rand() % ((len - cnt) + 1);
```
I think this still allows weight to be zero.  The +1 needs to be outside, since rand() can return 0.

shefty commented:
Thanks - nice addition!

aingerson commented:
Updated - the real equation this time!
